packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+      return null;
+    }
+    final ByteData encoded = data.buffer.asByteData(data.offsetInBytes, data.lengthInBytes);
+    return const StandardMessageCodec().decodeMessage(encoded) as Map<dynamic, dynamic>;
+  }
+
+  /// Called by the [RestorationManager] on itself to serialized the
+  /// restoration `data` before sending it to the engine via [sendToEngine].
+  ///
+  /// See also:
+  ///
+  ///  * [decodeRestorationData], which is the opposite of this method.
+  @protected
+  Uint8List encodeRestorationData(Map<dynamic, dynamic> data) {
+    final ByteData encoded = const StandardMessageCodec().encodeMessage(data);
+    return encoded.buffer.asUint8List(encoded.offsetInBytes, encoded.lengthInBytes);
+  }
+
+  bool _debugDoingUpdate = false;
+  bool _postFrameScheduled = false;
+
+  final Set<VoidCallback> _finalizers = <VoidCallback>{};
+
+  /// Schedules a post frame callback to serialize the current hierarchy of
+  /// restoration buckets and send it over to the engine.
+  ///
+  /// Called by [RestorationBucket]s whenever the data stored in them has
+  /// changed or the shape of the bucket hierarchy has been modified.
+  ///
+  /// The task will execute at the end of the current frame or - if currently
+  /// not producing a frame - at the end of the next frame. Just before the
+  /// hierarchy is serialized and send over to the engine, the optionally
+  /// provided `finalizer` is called. The callback may be used to check the
+  /// integrity of the data in a given bucket. It is illegal to call
+  /// [scheduleSerialization] from within the `finalizer` callback.
+  void scheduleSerialization({VoidCallback finalizer}) {
+    // The concept of `finalizer` callbacks is necessary to properly use
+    // restoration buckets in the [Widget] tree because of the way widgets are
+    // disposed during the build phase: Flutter delays the disposal of widgets
+    // until the very end of the build phase. When during a build widget A is
+    // getting replaced by widget B, B will build before A is disposed. This
+    // causes a problem if both widgets want to claim a child restoration bucket
+    // with ID x. In theory, this should be allowed because since A is getting
+    // replaced by B, A will give up the bucket in its dispose method so that B
+    // can claim a bucket with the same ID when it builds. However, in reality B
+    // will claim the bucket before A has released it, which means for a very
+    // short time (until A is actually getting disposed) there will be two child
+    // buckets under the same restoration ID, which is illegal. To get around
+    // this problem, restoration buckets do actually allow claiming child
+    // buckets for restoration IDs that have already been claimed in the hopes
+    // that the original owner will give up the bucket before the restoration
+    // data is finalized and sent to the engine at the end of the frame. They
+    // check whether this hope became reality in a finalizer callback.
```

Comment:
What if we made these more like streams with multiple subscribers? Would we still have this problem?

Alternatively, could we just have the framework "automatically" hand over ownership from A to B?